### DISCUSSION TITLE
alternatives to minimumNumberOfCalls

### DISCUSSION THREAD

**Discussion by gibffe** [resilience4j/resilience4j#1808] [Ideas]
Hello my friends - long time no see.
I've been using r4j for a while on various projects, and it has been humming away.

An idea recently came up which I would like to run past you.

Some of the backend systems I've been working on are experiencing substantial "day-night" cycle throughput amplitude changes, and I am starting to think `minimumNumberOfCalls` is no longer adequate - at peak throughput, the CB will trip almost immediately.

I wonder if an alternative setting, expressed as a % of the window size, would make more sense ? 
This could allow me to say collect statistics for a minimum of X seconds (X being a % of window size) regardless of throughput.

**Comment by RobWin** [comment]
Hi,
could you explain it in more detail?
I'm not sure if I understood the algorithm

**Comment by gibffe** [comment]
Hi Robert

with example core configuration:
```
sliding window type: TIME_BASED
sliding window size: 120s
minimum number of calls: 5000
```

at `tps=160` this circuit will have captured enough data to calculate error rate in ~ `33` second (27% of window size)

at `tps=500` this circuit will be ready to trip in ~ `10` seconds (8% of window size)

at `tps=800` this circuit will be ready to trip in ~ `6` seconds (5% of window size)

you see where this is going, window size becomes less and less relevant as tps increases

What I am looking for instead, is making the `minimumNumberOfCalls` adaptive, and a function of sliding window size.

Rather than having to hardcode it up front, I would like to have it frequently recalculated - such that it can adapt to load as it changes throughout the day:

```
if my 120s sliding window currently "captures" ~ 96000 requests
I can have my  minimumNumberOfCalls expressed as say 20% of that value - and it 
will be relevant under current conditions
```


Also, I would like to have an option of not specifying it at all, and always use full window size worth of data before calculating the thresholds.

**Reply by RobWin** [reply]
So the algorithm would calculate the `minimumNumberOfCalls` in percentage based on the number of recorded calls of the previous sliding window?

**Reply by gibffe** [reply]
Since this is a sliding window, I wasn't sure we can talk about "previous" window, but yes, that is exactly what I am thinking :)

**Reply by gibffe** [reply]
happy to look into a POC pull request for this

**Reply by RobWin** [reply]
Yes, I won't have time to do it by my own.
But I happy to assist you when you have questions.
We also have a slack channel
